src/script/mod.rs

```diff
@@ -115,6 +116,125 @@ impl From<opcode::Error> for Error {
     }
 }
 
+/// Evaluation functions for script components.
+pub trait Evaluable {
+    /// Get the byte length of this script sig.
+    fn byte_len(&self) -> usize;
+
+    /// Convert a sequence of `Opcode`s to the bytes that would be included in a transaction.
+    fn to_bytes(&self) -> Vec<u8>;
+
+    /// Evaluate this script component.
+    fn eval(
+        &self,
+        flags: interpreter::Flags,
+        checker: &dyn interpreter::SignatureChecker,
+        stack: interpreter::Stack<Vec<u8>>,
+    ) -> Result<interpreter::Stack<Vec<u8>>, Error>;
+
+    /// Returns true iff this script is P2SH.
+    fn is_pay_to_script_hash(&self) -> bool;
+
+    /// Called by `IsStandardTx` and P2SH/BIP62 VerifyScript (which makes it consensus-critical).
+    fn is_push_only(&self) -> bool;
+}
+
+/// A script component is used as either the script sig or script pubkey in a script. Depending on
+/// `T` (which generally needs an `opcode::Evaluable` instance), it has different properties:
+///
+/// - `PossiblyBad` – used for scripts that need to go through consensus-compatible verification
+///   (that is, read from the chain)
+/// - `Opcode` – used for script pubkey and non-push-only script sigs that are authored to be placed
+///   on the chain
+/// - `PushValue` – used for push-only script sigs that are authored to be placed on the chain
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct Component<T>(pub Vec<T>);
+
+/// A script sig has only `PushValue` elements.
+pub type Sig = Component<PushValue>;
+
+/// A script pubkey has any `Opcode`.
+pub type PubKey = Component<Opcode>;
+
+impl<T: opcode::Evaluable> Component<T> {
+    /// This parses an entire script. This is stricter than the incremental parsing that is done
+    /// during `verify_script`, because it fails on unknown opcodes no matter where they occur.
+    /// I.e., this is useful for validating and analyzing scripts before they are put into a
+    /// transaction, but not for scripts that are read from the chain, because it may fail on valid
+    /// scripts.
+    pub fn parse(raw_script: &Code) -> Result<Self, Error> {
```

Comment:
If `PossiblyBad` is added to the public API, then I think this documentation would be incorrect for it (because it wouldn't fail on unknown opcodes, just on incorrect structural encoding).

Ideally we'd have two parsing methods: `Component<PossiblyBad>::parse_lax` which only performs structural checks, and `Component<not PossiblyBad>::parse_strict` which does what this parser does. Given the lack of time to engineer that solution (which would likely add complexity due to needing an additional trait layer to express the `not PossiblyBad`), that is explicitly deferred to a future issue.

For this PR just alter the documentation to clarify that it is only stricter when `T` is not `PossiblyBad`.